commit c7319f6c3eb5b4636fa525e5379ab3979d426d96
Author: Lekensteyn <lekensteyn@gmail.com>
Date:   Thu Jan 26 22:51:43 2012 +0100

    Ignore empty KernelDriver values

diff --git a/src/bbconfig.c b/src/bbconfig.c
index f68290c..f583344 100644
--- a/src/bbconfig.c
+++ b/src/bbconfig.c
@@ -407,7 +407,14 @@ void bbconfig_parse_conf_driver(GKeyFile *bbcfg, char *driver) {
 
   key = "KernelDriver";
   if (g_key_file_has_key(bbcfg, section, key, NULL)) {
-    free_and_set_value(&bb_config.module_name, g_key_file_get_string(bbcfg, section, key, NULL));
+    char *module_name = g_key_file_get_string(bbcfg, section, key, NULL);
+    /* if KernelDriver is empty, the default behavior is to copy Driver which
+     * is done in driver_detect() */
+    if (*module_name) {
+      free_and_set_value(&bb_config.module_name, module_name);
+    } else {
+      g_free(module_name);
+    }
   }
   key = "LibraryPath";
   if (g_key_file_has_key(bbcfg, section, key, NULL)) {
