diff --git a/libglfork.cpp b/libglfork.cpp
index fa70f0e..e0667d0 100644
--- a/libglfork.cpp
+++ b/libglfork.cpp
@@ -37,8 +37,22 @@ struct CapturedFns {
   CapturedFns(const char *lib)
   {
     die_if(!lib || lib[0] != '/', "need absolute library path: %s\n", lib);
-    handle = dlopen(lib, RTLD_LAZY);
-    die_if(!handle, "%s\n", dlerror());
+    char *paths = strdup(lib);
+    char *p = paths, *p_next, *last_err;
+    do {
+      p_next = strchr(p, ':');
+      if (p_next) {
+        *p_next = 0;
+        p_next++;
+      }
+      handle = dlopen(p, RTLD_LAZY);
+      if (handle)
+        break;
+      last_err = dlerror();
+      die_if(strncmp(p, last_err, strlen(p)), "Failed to load %s: %s\n", p, last_err);
+    } while ((p = p_next) && *p);
+    free(paths);
+    die_if(!handle, "%s\n", last_err);
 #define DEF_GLX_PROTO(ret, name, args, ...) name = (ret (*) args)dlsym(handle, #name);
 #include "glx-reimpl.def"
 #include "glx-dpyredir.def"
