Description: use all variables containing hardening flags
Forwarded: https://github.com/amonakov/primus/pull/148
Author: Andreas Beckmann <anbe@debian.org>
Last-Update: 2014-10-25

--- a/Makefile
+++ b/Makefile
@@ -30,4 +30,4 @@
 
 $(LIBDIR)/libGL.so.1: libglfork.cpp
 	mkdir -p $(LIBDIR)
-	$(CXX) $(CXXFLAGS) -fvisibility=hidden -fPIC -shared -Wl,-Bsymbolic -o $@ $< -lX11 -lpthread -lrt
+	$(CXX) $(CPPFLAGS) $(CXXFLAGS) -fvisibility=hidden -fPIC -shared -Wl,-Bsymbolic -o $@ $< $(LDFLAGS) -lX11 -lpthread -lrt
